tests/cover/test_type_lookup.py

```diff
@@ -0,0 +1,140 @@
+            ex = thing(*([st.nothing()] * n)).example()
+            types_with_core_strat.add(type(ex))
+            break
+        except (TypeError, InvalidArgument):
+            continue
+
+
+@pytest.mark.parametrize('typ', sorted(types_with_core_strat, key=str))
+def test_resolve_core_strategies(typ):
+    @given(from_type(typ))
+    def inner(ex):
+        if PY2 and issubclass(typ, integer_types):
+            assert isinstance(ex, integer_types)
+        else:
+            assert isinstance(ex, typ)
+
+    inner()
+
+
+def test_lookup_knows_about_all_core_strategies():
+    cannot_lookup = types_with_core_strat - set(types.type_strategy_mapping())
+    assert not cannot_lookup
+
+
+def test_lookup_keys_are_types():
+    with pytest.raises(InvalidArgument):
+        st.register_type_strategy('int', st.integers())
+    assert 'int' not in types._global_type_to_strategy_lookup
+
+
+def test_lookup_values_are_strategies():
+    with pytest.raises(InvalidArgument):
+        st.register_type_strategy(int, 42)
+    assert 42 not in types._global_type_to_strategy_lookup.values()
+
+
+@pytest.mark.parametrize('typ', sorted(types_with_core_strat, key=str))
+def test_lookup_overrides_defaults(typ):
+    sentinel = object()
+    try:
+        st.register_type_strategy(typ, st.just(sentinel))
+        assert from_type(typ).example() is sentinel
+    finally:
+        st.register_type_strategy(typ, None)
+    assert from_type(typ).example() is not sentinel
+
+
+class ParentUnknownType(object):
+    pass
+
+
+class UnknownType(ParentUnknownType):
+    pass
+
+
+def test_custom_type_resolution():
+    with pytest.raises(ResolutionFailed):
+        from_type(UnknownType)
+    sentinel = object()
+    try:
+        st.register_type_strategy(UnknownType, st.just(sentinel))
+        assert from_type(UnknownType).example() is sentinel
+        # Also covered by registration of child class
+        assert from_type(ParentUnknownType).example() is sentinel
+    finally:
+        st.register_type_strategy(UnknownType, None)
+    with pytest.raises(ResolutionFailed):
+        from_type(UnknownType)
+
+
+def test_errors_if_generic_resolves_empty():
+    try:
+        st.register_type_strategy(UnknownType, lambda _: st.nothing())
+        with pytest.raises(ResolutionFailed):
+            from_type(UnknownType)
+        with pytest.raises(ResolutionFailed):
+            from_type(ParentUnknownType)
+    finally:
+        st.register_type_strategy(UnknownType, None)
+
+
+def test_cannot_register_empty():
+    # Cannot register and did not register
+    with pytest.raises(InvalidArgument):
+        st.register_type_strategy(UnknownType, st.nothing())
+    with pytest.raises(ResolutionFailed):
+        from_type(UnknownType)
+    assert UnknownType not in types._global_type_to_strategy_lookup
+
+
+def test_pulic_interface_works():
+    st.from_type(int).example()
+    with pytest.raises(InvalidArgument):
+        st.from_type('not a type or annotated function')
```

Comment:
This is actually counter to how all of the other strategies work, and I think should be fixed: Strategies are supposed to error when you first draw from them rather than at definition point, so that the errors are localized to the tests they're in rather than happening at module load time.